gobblin-data-management/src/main/java/org/apache/gobblin/data/management/source/LoopingDatasetFinderSource.java

```diff
@@ -170,19 +245,44 @@ public DeepIterator(Iterator<Dataset> baseIterator, String previousDatasetUrnWat
     @Override
     protected WorkUnit computeNext() {
       if (this.generatedWorkUnits >= this.maxWorkUnits) {
-        return endOfData();
+        if (retryWorkUnits.hasNext()) {
+          //For the case where retryWorkUnits.size() > maxWorkUnits, we need to keep track of retryWorkunits that aren't
+          //able to fit in a single execution, by committing their state thereby being recognizable in next execution.
+          WorkUnit addtionWorkUnit = retryWorkUnits.next();
+          TaskUtils.setTaskFactoryClass(addtionWorkUnit, RolloverTask.Factory.class);
+          return addtionWorkUnit;
+        } else {
+          return endOfData();
+        }
+      }
+
+      if (retryWorkUnits.hasNext()) {
+        WorkUnit workUnit = retryWorkUnits.next();
+        int previousRetryCount = workUnit.contains(ConfigurationKeys.TASK_RETRIES_KEY) ?
+            workUnit.getPropAsInt(ConfigurationKeys.TASK_RETRIES_KEY) + 1 : 1;
+        workUnit.setProp(ConfigurationKeys.TASK_RETRIES_KEY, previousRetryCount);
+        generatedWorkUnits++;
+        return workUnit;
       }
 
       while (this.baseIterator.hasNext() || this.currentPartitionIterator.hasNext()) {
         if (this.currentPartitionIterator != null && this.currentPartitionIterator.hasNext()) {
           PartitionableDataset.DatasetPartition partition = this.currentPartitionIterator.next();
+          // For partition appeared in retryUrns, just skip.
+          if (retryUrns.contains(partition.getUrn())) {
+            continue;
```

Comment:
I think we should add a `log.info()` message here for informational purposes and to help us debug in case we encounter issues.